### ISSUE TITLE
Allow remote storage to send data with warning and reflect warning in Prometheus API

### ISSUE THREAD

**Issue by mnottheone** [prometheus/prometheus#7161]
## Allow warning/error in response from remote storage along with data. Reflect Error/Warning in Prometheus final results.

Right now, Prometheus API shows warning when Querier get error [Corresponding issue-2573](https://github.com/prometheus/prometheus/issues/2573). My understanding is there is no way Remote Storage can send partial data will warnings and those warnings can reach user calling API or viewing Dashboards.<br>
**Use case. Why is this important?**
Usually Remote storage pull and aggregate raw data from multiple backends, e.g. for 1 promQL get metadata from some store X to figure out it needs N timeseries and then Query A timeseries from backend 1 and N-A from backend 2, join them and return back raw data. Let prometheus do final aggregations and return final result.

**Right now there is no Scope of partial failure for one remote store.**

Implementation : 
- Add warning/error to ReadResponse in remote.proto, to allow remote storage send data with warnings or errors. 
- Reflect Error/Warning in Prometheus Final results. Final warning could be simple comma separated concatenation of storage warnings.

**Comment by brian-brazil** [comment]
I'm not sure this is sane semantically. The reason the warning system works the way it does it that returning an arbitrary subset of data is not tractable for the end-user, so if there's an error from the remote endpoint we drop everything from that endpoint in order to keep things sane - in the same way that there's no such thing as a partial scrape.
Put another way, if a remote everything returns a warning then we'd also need to drop everything as we that's still arbitrarily partial data so this wouldn't gain us anything over returning an error.

**Comment by roidelapluie** [comment]
cc @csmarchbanks @cstyan, do you have some input for this issue?

**Comment by bwplotka** [comment]
I would be happy to accept PR for this, we do exactly this in Thanos Store API which is exactly remote read but in gRPC, see: https://github.com/thanos-io/thanos/blob/20004510b052923ff1e03f1d6017cd71eda423f0/pkg/store/storepb/rpc.proto#L124

```
message SeriesResponse {
  oneof result {
    /// series contains 1 response series. The series labels are sorted by name.
    Series series = 1;

    /// warning is considered an information piece in place of series for warning purposes.
    /// It is used to warn store API user about suspicious cases or partial response (if enabled).
    string warning = 2;

    /// hints is an opaque data structure that can be used to carry additional information from
    /// the store. The content of this field and whether it's supported depends on the
    /// implementation of a specific store. It's also implementation specific if it's allowed that
    /// multiple SeriesResponse frames contain hints for a single Series() request and how should they
    /// be handled in such case (ie. merged vs keep the first/last one).
    google.protobuf.Any hints = 3;
  }
}
```

The question is to @mnottheone is it still needed? Does any other user need this?

> Put another way, if a remote everything returns a warning then we'd also need to drop everything as that's still arbitrarily partial data so this wouldn't gain us anything over returning an error.

I think this would be case of warning without drop of such remote endpoint. So essentially user would be able to artificially create partial response situation by retuning OK even on errors and use warnings for error handling. Question is @brian-brazil do we want this ;p

**Comment by brian-brazil** [comment]
> Question is @brian-brazil do we want this ;p

Indeed, is there a sane setup where this would even come up? I'm not aware of one currently, and this seems like quite a bit of a foot-gun.

**Comment by mnottheone** [comment]
@bwplotka , Indeed Thanos was my motivation behind this proposal. When I checked other integrations like M3DB and [cortex](https://cortexmetrics.io/docs/architecture/), found out they have their own querier/cordinator and bypass prometheus API or leverage gRPC like Thanos. 

I will still like to have this. Fun fact I needed hint also and was passing hint via label and handling in my adapter.

At the same time I understand what @brian-brazil mean by foot-gun. When it comes to alerting/recording partial response can lead to false alerts/data. It needs to configurable in Prometheus instance what will be its behavior given warning. I think same is case with Thanos and we can default to warning is failure for remote storage.

>Put another way, if a remote everything returns a warning then we'd also need to drop everything as that's still arbitrarily partial data so this wouldn't gain us anything over returning an error.

Partial data can be useful depends on how partial it is: consider cases like
- Adapter internally querying 1000 and failed on 1.  
- Adapter querying data from different store divided based on time recent 3hr, 3h to 7d, and everything else. User might be okay with data not available for time beyond 7 days.
- Adapter merging data from different availability zone and failed for one. Slightly weaker case, I can think of suggestions myself.

**Comment by brian-brazil** [comment]
> Thanos was my motivation behind this proposal

Thanos doesn't use the Prometheus query APIs though, it goes direct to the remote read endpoint so this change wouldn't make a difference.

Can you give a better example of the use case where this is coming up?

**Comment by krajorama** [comment]
Hello from the bug scrub, there was no movement on this issue in the past 4 years. We're keeping the issue open for now as the remote read protobuf still does not have partial data with warnings, on the other hand we are not seeing this specific use case anymore since remote read is a less popular solution. Please feel free to follow up.
